com.unity.mobile.notifications/Runtime/Android/Plugins/com/unity/androidnotifications/UnityNotificationUtilities.java

```diff
@@ -510,46 +511,72 @@ private static <T extends Parcelable> T deserializeParcelable(DataInputStream in
+    //   * class specified in meta-data key custom_notification_android_activity
+    //   * the only enabled activity with name ending in either .UnityPlayerActivity or .UnityPlayerGameActivity
+    //   * the only enabled activity in the package
+    protected static Class<?> getOpenAppActivity(Context context) {
         try {
-            ai = context.getPackageManager().getApplicationInfo(context.getPackageName(), PackageManager.GET_META_DATA);
-        } catch (PackageManager.NameNotFoundException e) {
-            e.printStackTrace();
-        }
-        Bundle bundle = ai.metaData;
-
-        String customActivityClassName = null;
-        Class activityClass = null;
-
-        if (bundle.containsKey("custom_notification_android_activity")) {
-            customActivityClassName = bundle.getString("custom_notification_android_activity");
+            PackageManager pm = context.getPackageManager();
+            ApplicationInfo ai = pm.getApplicationInfo(context.getPackageName(), PackageManager.GET_META_DATA);
+            Bundle bundle = ai.metaData;
 
-            try {
-                activityClass = Class.forName(customActivityClassName);
-            } catch (ClassNotFoundException ignored) {
-                ;
+            if (bundle.containsKey("custom_notification_android_activity")) {
+                try {
+                    return Class.forName(bundle.getString("custom_notification_android_activity"));
+                } catch (ClassNotFoundException e) {
+                    Log.e(TAG_UNITY, "Specified activity class for notifications not found: " + e.getMessage());
+                }
             }
-        }
 
-        if (activityClass == null && fallbackToDefault) {
             Log.w(TAG_UNITY, "No custom_notification_android_activity found, attempting to find app activity class");
 
-            String classToFind = "com.unity3d.player.UnityPlayerActivity";
-            try {
-                return Class.forName(classToFind);
-            } catch (ClassNotFoundException ignored) {
-                Log.w(TAG_UNITY, String.format("Attempting to find : %s, failed!", classToFind));
-                classToFind = String.format("%s.UnityPlayerActivity", context.getPackageName());
-                try {
-                    return Class.forName(classToFind);
-                } catch (ClassNotFoundException ignored1) {
-                    Log.w(TAG_UNITY, String.format("Attempting to find class based on package name: %s, failed!", classToFind));
+            ActivityInfo[] aInfo = pm.getPackageInfo(context.getPackageName(), PackageManager.GET_ACTIVITIES).activities;
+            if (aInfo != null) {
+                String activityClassName = null;
+                String conflictingActivity = null;
+                boolean activityConflict = false;
+                for (ActivityInfo info : aInfo) {
+                    // activity alias not supported
+                    if (info.enabled && info.targetActivity == null) {
```

Comment:
Won't that only spread thread more vertically and still have the same problem?

Reply:
I was think something like this, this will also makes easier code to expand in case we'll have more activities in the future:

```
            ActivityInfo[] aInfo = pm.getPackageInfo(context.getPackageName(), PackageManager.GET_ACTIVITIES).activities;
            if (aInfo == null)
                return null;

            List<String> validActivities = new ArrayList<String>();
            for (ActivityInfo info : aInfo) {
                if (!info.enabled || info.targetActivity != null)
                    continue;

                if (info.name.endsWith(".UnityPlayerActivity") ||
                    info.name.endsWith(".UnityPlayerGameActivity"))
                    validActivities.add(info.name);
            }

            if (validActivities.size() > 0)
            {
                Log.e(TAG_UNITY, String.format("Multiple choices for activity for notifications: %s", String.join(", ", validActivities)));
                return null;
            } else if (validActivities.size() == 0)
            {
                Log.e(TAG_UNITY, "Activity class for notifications not found");
                return null;
            }
            else
            {
                return Class.forName(validActivities.get(0));
            }

        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        } catch (ClassNotFoundException e) {
            Log.e(TAG_UNITY, "Failed to find activity class: " + e.getMessage());
        }

        return null;
```

Reply:
This is also not correct :)
Any activity is fine as long as there is no .UnityPlayer*Activity. Will adding that extra logic still make it smaller and clearer?
Note, that this code runs when app is in background, so avoiding extra allocations and extra time spent is desirable. Not super crucial, but it should be a significant benefit in code clarity.

Reply:
Not sure if extra allocations is a problem here, but you current code is barely readable, please refactor :)